File: kotlin/exercises/src/main/java/io/github/ajoz/workshop/fp/java/part_3/solutions/exercise_1/Exercise1.java
package io.github.ajoz.workshop.fp.java.part_3.solutions.exercise_1;

import io.github.ajoz.workshop.fp.java.tools.Effect;
import io.github.ajoz.workshop.fp.java.tools.Supplier;

enum EnumBoolean {
    TRUE() {
        @Override
        public EnumBoolean and(final EnumBoolean other) {
            // if this is TRUE then `and` relies on other
            return other;
        }

        @Override
        public EnumBoolean or(final EnumBoolean other) {
            // if this is TRUE then `or` does not depend on other
            return this;
        }

        @Override
        public EnumBoolean not() {
            return FALSE;
        }

        @Override
        public EnumBoolean ifTrue(final Effect effect) {
            effect.perform();
            return this;
        }

        @Override
        public EnumBoolean ifFalse(final Effect effect) {
            return this; //pass through
        }

        @Override
        public <A> A match(final Supplier<A> ifTrue,
                           final Supplier<A> ifFalse) {
            return ifTrue.get();
        }
    },
    FALSE() {
        @Override
        public EnumBoolean and(final EnumBoolean other) {
            // if this is FALSE then `and` does not depend on other
            return this;
        }

        @Override
        public EnumBoolean or(final EnumBoolean other) {
            // if this is FALSE then `or` relies on other
            return other;
        }

        @Override
        public EnumBoolean not() {
            return TRUE;
        }

        @Override
        public EnumBoolean ifTrue(final Effect effect) {
            return this;
        }

        @Override
        public EnumBoolean ifFalse(final Effect effect) {
            effect.perform();
            return this;
        }

        @Override
        public <A> A match(final Supplier<A> ifTrue,
                           final Supplier<A> ifFalse) {
            return ifFalse.get();
        }
    };

    public abstract EnumBoolean and(final EnumBoolean other);

    public abstract EnumBoolean or(final EnumBoolean other);

    public abstract EnumBoolean not();

    public abstract EnumBoolean ifTrue(final Effect effect);

    public abstract EnumBoolean ifFalse(final Effect effect);

    public abstract <A> A match(final Supplier<A> ifTrue,
                                final Supplier<A> ifFalse);
}

@SuppressWarnings("ConstantConditions")
final class Exercise1 {
    public static void main(final String[] args) {
        // simple boolean algebra:
        // and:
        System.out.println(EnumBoolean.TRUE.and(EnumBoolean.TRUE));
        System.out.println(EnumBoolean.TRUE.and(EnumBoolean.FALSE));
        System.out.println(EnumBoolean.FALSE.and(EnumBoolean.FALSE));
        System.out.println(EnumBoolean.FALSE.and(EnumBoolean.TRUE));

        // or:
        System.out.println(EnumBoolean.TRUE.or(EnumBoolean.FALSE));
        System.out.println(EnumBoolean.FALSE.or(EnumBoolean.TRUE));
        System.out.println(EnumBoolean.FALSE.or(EnumBoolean.FALSE));
        System.out.println(EnumBoolean.TRUE.or(EnumBoolean.TRUE));

        // not:
        System.out.println(EnumBoolean.TRUE.not());
        System.out.println(EnumBoolean.FALSE.not());

        // conditions as methods:
        EnumBoolean.TRUE
                .ifTrue(() -> System.out.println("EnumTrue is EnumTrue!"))
                .ifFalse(() -> System.out.println("EnumTrue is EnumFalse?"));

        EnumBoolean.FALSE
                .ifFalse(() -> System.out.println("EnumFalse is EnumFalse!"))
                .ifTrue(() -> System.out.println("EnumFalse is EnumTrue?"));

        // conditions as expressions:
        final String trueMessage =
                EnumBoolean.TRUE
                        .match(
                                () -> "Matching EnumTrue to EnumTrue!",
                                () -> "Matching EnumTrue to EnumFalse?"
                        );
        System.out.println(trueMessage);

        final String falseMessage =
                EnumBoolean.FALSE
                        .match(
                                () -> "Matching EnumFalse to EnumTrue?",
                                () -> "Matching EnumFalse to EnumFalse!"
                        );
        System.out.println(falseMessage);

        // using the switch:
        final EnumBoolean bool = EnumBoolean.TRUE;
        switch (bool) {
            case FALSE:
                System.out.println("Switch FALSE");
                break;
            case TRUE:
                System.out.println("Switch TRUE");
                break;
        }
    }
}

File: kotlin/exercises/src/main/kotlin/io/github/ajoz/workshop/fp/kotlin/part_3/solutions/exercise_1/Exercise1.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_3.solutions.exercise_1

import io.github.ajoz.workshop.fp.kotlin.tools.also

enum class EnumBoolean {
    TRUE {
        override fun and(other: EnumBoolean) = other
        override fun or(other: EnumBoolean) = this
        override operator fun not() = FALSE

        // I've added a override to also to be able to pass a () -> Unit
        override fun ifTrue(effect: () -> Unit) = also(effect)

        override fun ifFalse(effect: () -> Unit) = this
        override fun <A> match(ifTrue: () -> A, ifFalse: () -> A): A = ifTrue()
    },
    FALSE {
        override fun and(other: EnumBoolean) = this
        override fun or(other: EnumBoolean) = other
        override operator fun not() = TRUE

        override fun ifTrue(effect: () -> Unit) = this
        override fun ifFalse(effect: () -> Unit) = also(effect)

        override fun <A> match(ifTrue: () -> A, ifFalse: () -> A) = ifFalse()
    };

    abstract infix fun and(other: EnumBoolean): EnumBoolean
    abstract infix fun or(other: EnumBoolean): EnumBoolean
    abstract operator fun not(): EnumBoolean

    abstract fun ifTrue(effect: () -> Unit): EnumBoolean
    abstract fun ifFalse(effect: () -> Unit): EnumBoolean

    abstract fun <A> match(ifTrue: () -> A, ifFalse: () -> A): A
}

fun main(args: Array<String>) {
    // simple boolean algebra:
    // and:
    println(EnumBoolean.TRUE and EnumBoolean.TRUE)
    println(EnumBoolean.TRUE and EnumBoolean.FALSE)
    println(EnumBoolean.FALSE and EnumBoolean.FALSE)
    println(EnumBoolean.FALSE and EnumBoolean.TRUE)

    // or:
    println(EnumBoolean.TRUE or EnumBoolean.FALSE)
    println(EnumBoolean.FALSE or EnumBoolean.TRUE)
    println(EnumBoolean.FALSE or EnumBoolean.FALSE)
    println(EnumBoolean.TRUE or EnumBoolean.TRUE)

    // not:
    println(EnumBoolean.TRUE.not())
    println(EnumBoolean.FALSE.not())

    // conditions as methods:
    EnumBoolean.TRUE
            .ifTrue { println("EnumTrue is EnumTrue!") }
            .ifFalse { println("EnumTrue is EnumFalse?") }

    EnumBoolean.FALSE
            .ifFalse { println("EnumFalse is EnumFalse!") }
            .ifTrue { println("EnumFalse is EnumTrue?") }

    // conditions as expressions:
    val trueMessage = EnumBoolean.TRUE
            .match(
                    { "Matching EnumTrue to EnumTrue!" },
                    { "Matching EnumTrue to EnumFalse?" }
            )
    println(trueMessage)

    val falseMessage = EnumBoolean.FALSE
            .match(
                    { "Matching EnumFalse to EnumTrue?" },
                    { "Matching EnumFalse to EnumFalse!" }
            )
    println(falseMessage)

    // using when:
    val bool = EnumBoolean.TRUE
    when (bool) {
        EnumBoolean.FALSE -> println("When FALSE")
        EnumBoolean.TRUE -> println("When TRUE")
    }
}

